File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/EmailSample/src/com/hp/workpath/sample/emailsample/filebrowser/FileLoader.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.emailsample.filebrowser;

import android.content.Context;

import androidx.loader.content.AsyncTaskLoader;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class FileLoader extends AsyncTaskLoader<List<File>> {

    private List<File> mData;
    private String mPath;

    public FileLoader(Context context, String path) {
        super(context);
        this.mPath = path;
    }

    @Override
    public List<File> loadInBackground() {
        ArrayList<File> list = new ArrayList<File>();

        final File pathDir = new File(mPath);
        final File[] dirs = pathDir.listFiles(FileUtils.DIR_FILTER);
        if (dirs != null) {
            Arrays.sort(dirs, FileUtils.COMPARATOR);
            Collections.addAll(list, dirs);
        }

        final File[] files = pathDir.listFiles(FileUtils.FILE_FILTER);
        if (files != null) {
            Arrays.sort(files, FileUtils.COMPARATOR);
            Collections.addAll(list, files);
        }
        return list;
    }

    @Override
    public void deliverResult(List<File> data) {
        if (isReset()) {
            return;
        }

        mData = data;

        if (isStarted()) {
            super.deliverResult(data);
        }
    }

    @Override
    protected void onStartLoading() {
        if (mData != null)
            deliverResult(mData);

        if (takeContentChanged() || mData == null)
            forceLoad();
    }

    @Override
    protected void onStopLoading() {
        cancelLoad();
    }

    @Override
    protected void onReset() {
        onStopLoading();
        mData = null;
    }

    @Override
    public void onCanceled(List<File> data) {
        super.onCanceled(data);
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/EmailSample/src/com/hp/workpath/sample/emailsample/filebrowser/FileLoader.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.emailsample.filebrowser

import android.content.Context
import androidx.loader.content.AsyncTaskLoader
import java.io.File
import java.util.Arrays
import java.util.Collections

class FileLoader(context: Context, private val mPath: String) : AsyncTaskLoader<List<File>>(context) {
    private var mData: List<File>? = null
    override fun loadInBackground(): List<File> {
        val list = ArrayList<File>()
        val pathDir = File(mPath)
        val dirs = pathDir.listFiles(FileUtils.DIR_FILTER)
        if (dirs != null) {
            Arrays.sort(dirs, FileUtils.COMPARATOR)
            Collections.addAll(list, *dirs)
        }
        val files = pathDir.listFiles(FileUtils.FILE_FILTER)
        if (files != null) {
            Arrays.sort(files, FileUtils.COMPARATOR)
            Collections.addAll(list, *files)
        }
        return list
    }

    override fun deliverResult(data: List<File>?) {
        if (isReset) {
            return
        }
        mData = data
        if (isStarted) {
            super.deliverResult(data)
        }
    }

    override fun onStartLoading() {
        if (mData != null) deliverResult(mData)
        if (takeContentChanged() || mData == null) forceLoad()
    }

    override fun onStopLoading() {
        cancelLoad()
    }

    override fun onReset() {
        onStopLoading()
        mData = null
    }

    override fun onCanceled(data: List<File>?) {
        super.onCanceled(data)
    }

}